contracts/generator_controller/src/contract.rs

```diff
@@ -0,0 +1,451 @@
+                    v.owner = new_owner;
+                    Ok(v)
+                })?;
+
+                Ok(())
+            })
+            .map_err(|e| e.into())
+        }
+    }
+}
+
+/// ## Description
+/// The function checks that:
+/// * the user voting power is > 0,
+/// * user didn't vote for last 10 days,
+/// * all pool addresses are valid,
+/// * 'votes' vector doesn't contain duplicated pool addresses,
+/// * sum of all BPS values <= 10000.
+///
+/// The function cancels changes applied by previous votes and apply new votes for the next period.
+/// New vote parameters are saved in [`USER_INFO`].  
+///
+/// The function returns [`Response`] in case of success or [`ContractError`] in case of errors.
+///
+/// ## Params
+/// * **deps** is an object of type [`DepsMut`].
+///
+/// * **env** is an object of type [`Env`].
+///
+/// * **info** is an object of type [`MessageInfo`].
+///
+/// * **votes** is a vector of pairs ([`String`], [`u16`]).
+/// Pair consists of pool address and percentage of user's voting power for a given pool.
+/// Percentage should be in BPS form.
+fn handle_vote(
+    mut deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    votes: Vec<(String, u16)>,
+) -> ExecuteResult {
+    let user = info.sender;
+    let block_period = get_period(env.block.time.seconds())?;
+    let escrow_addr = CONFIG.load(deps.storage)?.escrow_addr;
+    let mut user_vp = get_voting_power(deps.querier, &escrow_addr, &user)?;
+
+    if user_vp.is_zero() {
+        return Err(ContractError::ZeroVotingPower {});
+    }
+
+    let user_info = USER_INFO.may_load(deps.storage, &user)?.unwrap_or_default();
+    // Does the user eligible to vote again?
+    if env.block.time.seconds() - user_info.vote_ts < VOTE_COOLDOWN {
+        return Err(ContractError::CooldownError(VOTE_COOLDOWN / DAY));
+    }
+
+    // Check duplicated votes
+    let addrs_set = votes
+        .iter()
+        .cloned()
+        .map(|(addr, _)| addr)
+        .collect::<HashSet<_>>();
+    if votes.len() != addrs_set.len() {
+        return Err(ContractError::DuplicatedPools {});
+    }
+
+    // Validating addrs and bps
+    let votes = votes
+        .into_iter()
+        .map(|(addr, bps)| {
+            let addr = addr_validate_to_lower(deps.api, &addr)?;
+            let bps: BasicPoints = bps.try_into()?;
+            Ok((addr, bps))
+        })
+        .collect::<Result<Vec<_>, ContractError>>()?;
+
+    // Check the bps sum is within the limit
+    votes
+        .iter()
+        .try_fold(BasicPoints::default(), |acc, (_, bps)| {
+            acc.checked_add(*bps)
+        })?;
+
+    let user_last_vote_period = get_period(user_info.vote_ts).unwrap_or(block_period);
+
+    // Calculate voting power before changes
+    let old_vp_at_period = calc_voting_power(
+        user_info.slope,
+        user_info.voting_power,
+        user_last_vote_period,
+        block_period,
+    );
+
+    if user_info.lock_end > block_period {
```

Comment:
I think we should use `>=` comparison here because when user_info.lock_end equals to the current block_period we need to recalculate pools slope changes at the next block due to new pool votes

Reply:
Because all vote changes (cancelling prev. changes and appending new votes) are applied to the next period, a votes for the current period are already calculated before and they are considered immutable. This line checks whether the user's lock finishes in the next period and later or not. Thus we cancel user's changes in a mutable voted pool state at the next period.
If we set ">=" here we will cancel users changes again which will be already cancelled in the current period.